app/src/main/java/com/kevinlu/watstats/LoginActivity.java

```diff
@@ -0,0 +1,187 @@
+            public void onReceivedError(WebView view, WebResourceRequest request,
+                                        WebResourceError error) {
+
+                Toast.makeText(getApplicationContext(),
+                        "WebView Error",
+                        Toast.LENGTH_SHORT).show();
+
+                super.onReceivedError(view, request, error);
+            }
+
+            @Override
+            public void onPageFinished(WebView view, String url) {
+                switch (url) {
+                    case "https://watcard.uwaterloo.ca/OneWeb/Account/LogOn#first":
+                        webDriver.loadUrl("javascript:(function(){" +
+                                String.format("document.getElementById('Account').value='%s';", id) +
+                                String.format("document.getElementById('Password').value='%s';", pwd) +
+                                "document.forms[1].submit();" +
+                                "})()");
+                        break;
+                    case "https://watcard.uwaterloo.ca/OneWeb/Account/LogOn":
+                        _loginButton.setEnabled(true);
+
+                        Toast.makeText(getBaseContext(), "Incorrect login details", Toast.LENGTH_LONG).show();
+                        progressDialog.dismiss();
+                        break;
+                    case "https://watcard.uwaterloo.ca/OneWeb/Account/Personal":
+                        Toast.makeText(getBaseContext(), "Logged in", Toast.LENGTH_LONG).show();
+                        _loginButton.setEnabled(true);
+
+                        Intent intent = new Intent(getApplicationContext(), MainActivity.class);
+                        startActivity(intent);
+                        progressDialog.dismiss();
+                        break;
+                }
+            }
+        });
+        webDriver.loadUrl("https://watcard.uwaterloo.ca/OneWeb/Account/LogOn#first");
+    }
+
+    @Override
+    public void onBackPressed() {
+        // disable going back to the MainActivity
+        moveTaskToBack(true);
+    }
+
+    public void onLoginSuccess() {
+        _loginButton.setEnabled(true);
+        //finish();
+    }
+
+    public void onLoginFailed() {
+        Toast.makeText(getBaseContext(), "Login failed", Toast.LENGTH_LONG).show();
+        _loginButton.setEnabled(true);
+    }
+
+    public boolean validate() {
+        boolean valid = true;
+
+        String watcard = _watcardText.getText().toString();
+        String password = _passwordText.getText().toString();
+
+        if (watcard.isEmpty()) {
+            _watcardText.setError("Enter your Watcard number");
+            valid = false;
+        } else {
+            _watcardText.setError(null);
+        }
+
+        if (password.isEmpty()) {
+            _passwordText.setError("Enter your Watcard password");
+            valid = false;
+        } else {
+            _passwordText.setError(null);
+        }
+
+        return valid;
+    }
```

Comment:
Instead of returning valid, it may be more readable to just return true or false at every instance where you assign a new value to it.

Reply:
This file was just a test, it will be deleted.